### PR TITLE
enh(javascript) Match numeric literals per ECMA-262 spec

### PR DESCRIPTION
**Pull request by gibson042**
https://tc39.es/ecma262/#sec-literals-numeric-literals
https://tc39.es/ecma262/#sec-additional-syntax-numeric-literals

### Changes
Brought JavaScript numeric literal variants in line with the [spec](https://tc39.es/ecma262/#sec-literals-numeric-literals), just like was done for Python in #2780 and Ruby in #2788.

### Checklist
- [x] Added markup tests
- [x] Updated the changelog at `CHANGES.md`
- [x] Added myself to `AUTHORS.txt`, under Contributors

### CODE REVIEW

File: src/languages/javascript.js [hunk 1]
```diff
@@ -63,29 +63,36 @@ export default function(hljs) {
...
+  // https://tc39.es/ecma262/#sec-literals-numeric-literals
+  const decimalDigits = '[0-9](_?[0-9])*';
+  const frac = `\\.(${decimalDigits})`;
+  // DecimalIntegerLiteral, including Annex B NonOctalDecimalIntegerLiteral
+  // https://tc39.es/ecma262/#sec-additional-syntax-numeric-literals
+  const decimalInteger = `0|[1-9](_?[0-9])*|0[0-7]*[89][0-9]*`;
   const NUMBER = {
     className: 'number',
     variants: [
-      { begin: nonDecimalLiterals('bB', '01') }, // Binary literals
```
Comment (joshgoebel) [inline src/languages/javascript.js:74]: I just wrote these beautiful abstractions not so long ago... 😐   Not saying change it, just shedding a quick tear.

File: src/languages/javascript.js [hunk 2]
```diff
@@ -63,29 +63,36 @@ export default function(hljs) {
...
+        `[eE][+-]?(${decimalDigits})\\b` },
+      { begin: `\\b(${decimalInteger})\\b((${frac})\\b|\\.)?|(${frac})\\b` },
+
+      // DecimalBigIntegerLiteral
+      { begin: `\\b(0|[1-9](_?[0-9])*)n\\b` },
+
+      // NonDecimalIntegerLiteral
+      { begin: "\\b0[bB][0-1](_?[0-1])*n?\\b" },
+      { begin: "\\b0[oO][0-7](_?[0-7])*n?\\b" },
+      { begin: "\\b0[xX][0-9a-fA-F](_?[0-9a-fA-F])*n?\\b" },
```
Comment (joshgoebel) [inline src/languages/javascript.js:87]: Who uses octal anyways?  Let's move hex and binary first in the sequencing... (perhaps a tiny perf improvement, but easy enough to do).

**Comment by joshgoebel** [comment]
@gibson042 Wow... only ~180 more languages to go! ;-)

**Review by joshgoebel — APPROVED** [review]

File: src/languages/javascript.js (hunk 2 above)
Comment (gibson042) [reply src/languages/javascript.js:87]: This sequence matches the [spec](https://www.ecma-international.org/ecma-262/11.0/index.html#prod-NonDecimalIntegerLiteral) and I'm not sure what reordering you want; can you clarify and/or update on your own?

**Comment by gibson042** [comment]
I'm definitely going for low-hanging fruit of languages with which I'm familiar and can refer to some kind of documentation, but may well hit some others in the C/Java family.

File: src/languages/javascript.js (hunk 2 above)
Comment (joshgoebel) [inline src/languages/javascript.js:87]:
```suggestion
      { begin: "\\b0[xX][0-9a-fA-F](_?[0-9a-fA-F])*n?\\b" },
      { begin: "\\b0[bB][0-1](_?[0-1])*n?\\b" },
      { begin: "\\b0[oO][0-7](_?[0-7])*n?\\b" },
```

**Review by joshgoebel — APPROVED** [review]

**Comment by joshgoebel** [comment]
@gibson042 Thanks for work on this!

**Pull request merged**